File: wear/src/main/java/com/windkracht8/musicplayer/Library.java
package com.windkracht8.musicplayer;

import android.content.ContentUris;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.Environment;
import android.provider.MediaStore;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.net.URI;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Objects;

public class Library{
    public static String exStorageDir;
    public final ArrayList<Track> tracks = new ArrayList<>();
    public final ArrayList<Artist> artists = new ArrayList<>();
    public final ArrayList<Album> albums = new ArrayList<>();

    public static int libraryScanVersion = 0;

    private final Player player = new Player();

    //TODO: toast errors
    public Library(){
        exStorageDir = Environment.getExternalStorageDirectory().toString();
    }
    public JSONArray getTracks(){
        JSONArray array = new JSONArray();
        for(Track track : tracks){
            array.put(track.toJson());
        }
        return array;
    }
    public void scanMediaStore(Main main){
        tracks.clear();
        artists.clear();
        albums.clear();

        String[] projection = new String[]{
                MediaStore.Audio.Media._ID,
                MediaStore.MediaColumns.DATA,
                MediaStore.Audio.Media.TITLE,
                MediaStore.Audio.Media.ARTIST,
                MediaStore.Audio.Media.ALBUM,
                MediaStore.Audio.Media.ALBUM_ARTIST,
                MediaStore.Audio.Media.CD_TRACK_NUMBER
        };

        try(Cursor cursor = main.getContentResolver().query(
                MediaStore.Audio.Media.EXTERNAL_CONTENT_URI,
                projection,
                null,
                null,
                null,
                null
        )) {
            if(cursor == null){
                Log.e(Main.LOG_TAG, "Library.rescan: Cursor is null");
                return;
            }
            int ID = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media._ID);
            int DATA = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DATA);
            int TITLE = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.TITLE);
            int ARTIST = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.ARTIST);
            int ALBUM = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.ALBUM);
            int ALBUM_ARTIST = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.ALBUM_ARTIST);
            int CD_TRACK_NUMBER = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.CD_TRACK_NUMBER);

            while(cursor.moveToNext()){
                Uri contentUri = ContentUris.withAppendedId(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, cursor.getLong(ID));
                new Track(main,
                        contentUri,
                        cursor.getString(DATA),
                        cursor.getString(TITLE),
                        cursor.getString(ARTIST),
                        cursor.getString(ALBUM),
                        cursor.getString(ALBUM_ARTIST),
                        cursor.getString(CD_TRACK_NUMBER)
                );
            }
        }
        Collections.sort(tracks);
        Collections.sort(artists);
        Collections.sort(albums);
        for(Artist artist : artists) artist.sort();
        for(Album album : albums) album.sort();

        libraryScanVersion++;
        if(libraryScanVersion == 1){
            main.handler_message.sendMessage(main.handler_message.obtainMessage(Main.MESSAGE_LIBRARY_READY));
            scanFiles(main);
        }
    }
    public void scanFiles(Main main){
        if(scanFiles(main, "")){
            scanMediaStore(main);
        }
    }
    public boolean scanFiles(Main main, String subdir){
        boolean addedTracks = false;
        String path = exStorageDir+"/Music"+subdir;
        File directory = new File(path);
        File[] files = directory.listFiles();
        if(files == null) return false;
        for(File file : files){
            if(file.isDirectory()){
                if(scanFiles(main, subdir+"/"+file.getName()))
                    addedTracks = true;
            }
            if(!file.getName().endsWith(".mp3")) continue;
            if(!trackExists(file.toURI())){
                player.load(main, Uri.fromFile(file));
                addedTracks = true;
            }
        }
        return addedTracks;
    }
    private boolean trackExists(URI uri){
        String path = uri.getPath().substring(exStorageDir.length()+1);
        for(Track track : tracks){
            if(track.path.equals(path)) return true;
        }
        return false;
    }
    public boolean ensurePath(String path){
        File file = new File(exStorageDir + "/" + path);
        try{
            if(file.exists()){
                Log.e(Main.LOG_TAG, "Library.ensurePath: path exists");
                return false;
            }

            File parent = file.getParentFile();
            if(parent == null){
                Log.e(Main.LOG_TAG, "Library.ensurePath: getParentFile == null");
                return false;
            }
            if(!parent.exists()){
                if(!parent.mkdirs()){
                    Log.e(Main.LOG_TAG, "Library.ensurePath: mkdirs");
                    return false;
                }
            }
            if(!file.createNewFile()){
                Log.e(Main.LOG_TAG, "Library.ensurePath: createNewFile");
                return false;
            }
        }catch(Exception e){
            Log.e(Main.LOG_TAG, "Library.ensurePath exception: " + e.getMessage());
            return false;
        }
        return true;
    }
    public void addFile(Main main, String path){
        File file = new File(exStorageDir + "/" + path);
        player.load(main, Uri.fromFile(file));
        scanMediaStore(main);
    }
    public String deleteFile(String path){
        File file = new File(exStorageDir + "/" + path);
        try{
            if(!file.exists()){
                Log.e(Main.LOG_TAG, "Library.deleteFile: path does not exists");
                return "OK";
            }
            if(!file.delete()){
                Log.e(Main.LOG_TAG, "Library.deleteFile: delete");
                return "delete failed";
            }
        }catch(Exception e){
            Log.e(Main.LOG_TAG, "Library.deleteFile exception: " + e.getMessage());
            return e.getMessage();
        }
        //TODO: remove from tracks, albums, artists, and reload MediaStore
        return "OK";
    }
    public class Track implements Comparable<Track>{
        public Uri uri;//content://media/external/audio/media/32
        public String path;//Music/Austrian Death Machine - Jingle Bells.mp3
        public String title;//Jingle Bells
        public Artist artist;
        public Album album;
        public String track_no;
        public Track(Context context, Uri uri, String path, String title, String artistName, String albumName, String albumArtist, String track_no){
            this.uri = uri;
            this.path = path.substring(exStorageDir.length()+1);
            this.title = title;
            this.track_no = track_no;

            artist = getArtistForNewTrack(this, artistName);
            album = getAlbumForNewTrack(context, this, albumName, albumArtist);
            artist.addAlbum(album);
            tracks.add(this);
        }
        public JSONObject toJson(){
            JSONObject object = new JSONObject();
            try{
                object.put("path", path);
            }catch(Exception e){
                Log.e(Main.LOG_TAG, "Library.Track.toJson: " + e.getMessage());
                //TODO: toast
            }
            return object;
        }
        @Override
        public int compareTo(Track track){
            int album = this.album.name.compareTo(track.album.name);
            if(album != 0) return album;
            int track_no = this.track_no.compareTo(track.track_no);
            if(track_no != 0) return track_no;
            return this.title.compareTo(track.title);
        }
    }
    public class Artist implements Comparable<Artist>{
        public String name;
        public final ArrayList<Track> artist_tracks = new ArrayList<>();
        public final ArrayList<Album> artist_albums = new ArrayList<>();
        public Artist(Track track, String artistName){
            name = artistName == null ? "<empty>" : artistName;
            artist_tracks.add(track);
            artists.add(this);
        }
        public void addAlbum(Album album){
            if(artist_albums.contains(album)) return;
            artist_albums.add(album);
        }
        @Override
        public int compareTo(Artist artist){
            return this.name.compareTo(artist.name);
        }
        public void sort(){
            Collections.sort(artist_tracks);
            Collections.sort(artist_albums);
        }
    }
    public Artist getArtistForNewTrack(Track track, String artistName){
        for(Artist artist : artists){
            if(artist.name.equals(artistName)){
                artist.artist_tracks.add(track);
                return artist;
            }
        }
        return new Artist(track, artistName);
    }
    public class Album implements Comparable<Album>{
        public String name;
        public String artist;
        public final ArrayList<Track> album_tracks = new ArrayList<>();
        public Album(Track track, String albumName, String albumArtist){
            name = albumName == null ? "<empty>" : albumName;
            artist = albumArtist;
            album_tracks.add(track);
            albums.add(this);
        }
        @Override
        public int compareTo(Album album){
            return this.name.compareTo(album.name);
        }
        public void sort(){
            Collections.sort(album_tracks);
        }
    }
    public Album getAlbumForNewTrack(Context context, Track track, String albumName, String albumArtist){
        for(Album album : albums){
            if(album.name.equals(albumName)){
                if(album.artist != null && !album.artist.equals(albumArtist)){
                    album.artist = context.getString(R.string.various_artists);
                }
                album.album_tracks.add(track);
                return album;
            }
        }
        return new Album(track, albumName, albumArtist);
    }
}

File: wear/src/main/java/data/Library.kt
/*
 * Copyright 2024-2026 Bart Vullings <dev@windkracht8.com>
 * This file is part of WearMusicPlayer
 * WearMusicPlayer is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 * WearMusicPlayer is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.windkracht8.wearmusicplayer.data

import android.content.ContentUris
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.media.MediaScannerConnection
import android.net.Uri
import android.os.Environment
import android.os.Process
import android.provider.MediaStore
import androidx.activity.result.IntentSenderRequest
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.setValue
import com.windkracht8.wearmusicplayer.CommsBT
import com.windkracht8.wearmusicplayer.Main
import com.windkracht8.wearmusicplayer.R
import com.windkracht8.wearmusicplayer.error
import com.windkracht8.wearmusicplayer.logD
import com.windkracht8.wearmusicplayer.logE
import com.windkracht8.wearmusicplayer.logI
import com.windkracht8.wearmusicplayer.runInBackground
import kotlinx.coroutines.flow.MutableSharedFlow
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

object Library {
	private val exStorageDir: String = Environment.getExternalStorageDirectory().toString()
	val musicDir: String = "$exStorageDir/Music/"
	val tracks = mutableStateListOf<Track>()
	var shuffleCounter by mutableIntStateOf(0)
	val artists = mutableStateListOf<Artist>()
	val albums = mutableStateListOf<Album>()
	val dirs = mutableStateListOf<Dir>()
	enum class Status { SCAN, READY, UPDATE }
	val status = MutableSharedFlow<Status>()
	private val projection: Array<String> = arrayOf(
		MediaStore.Audio.Media._ID,
		MediaStore.MediaColumns.DATA,
		MediaStore.Audio.Media.TITLE,
		MediaStore.Audio.Media.ARTIST,
		MediaStore.Audio.Media.ALBUM,
		MediaStore.Audio.Media.ALBUM_ARTIST,
		MediaStore.Audio.Media.CD_TRACK_NUMBER,
		MediaStore.Audio.Media.DISC_NUMBER
	)

	suspend fun scanMediaStore(context: Context) {
		logD { "Library.scanMediaStore" }
		status.emit(Status.SCAN)
		tracks.clear()
		artists.clear()
		albums.clear()
		dirs.clear()
		scanUri(context, MediaStore.Audio.Media.getContentUri(MediaStore.VOLUME_EXTERNAL))
		status.emit(Status.READY)
	}
	suspend fun scanFiles(context: Context) {
		logD { "Library.scanFiles" }
		status.emit(Status.SCAN)
		MediaScannerConnection.scanFile(
			context,
			arrayOf(exStorageDir),
			null
		) { _: String, _: Uri? -> runInBackground { scanMediaStore(context) } }
	}
	fun scanFile(context: Context, path: String) {
		logD { "Library.scanFile: $path" }
		MediaScannerConnection.scanFile(
			context,
			arrayOf("$musicDir$path"),
			null
		) { _: String, uri: Uri? ->
			if(uri == null) {
				logD { "Library.scanFile path does not exists" }
				removeTrack(path)
			} else {
				logD { "Library.scanFile path exists" }
				runInBackground {
					scanUri(context, uri)
					status.emit(Status.UPDATE)
				}
			}
		}
	}
	private suspend fun scanUri(context: Context, uri: Uri) {
		try {
			context.contentResolver.query(
				uri,
				projection,
				null,
				null,
				null,
				null
			).use { cursor ->
				//logD{"Library.scanMediaStore query done"}
				if(cursor == null) {
					logE("Library.scanUri: Cursor is null")
					error.emit(R.string.fail_scan_media)
					return
				}
				val id = cursor.getColumnIndex(MediaStore.Audio.Media._ID)
				val data = cursor.getColumnIndex(MediaStore.MediaColumns.DATA)
				val title = cursor.getColumnIndex(MediaStore.Audio.Media.TITLE)
				val artist = cursor.getColumnIndex(MediaStore.Audio.Media.ARTIST)
				val album = cursor.getColumnIndex(MediaStore.Audio.Media.ALBUM)
				val albumArtist = cursor.getColumnIndex(MediaStore.Audio.Media.ALBUM_ARTIST)
				val cdTrackNumber = cursor.getColumnIndex(MediaStore.Audio.Media.CD_TRACK_NUMBER)
				val discNumber = cursor.getColumnIndex(MediaStore.Audio.Media.DISC_NUMBER)
				while(cursor.moveToNext()) {
					val uri = ContentUris.withAppendedId(
						MediaStore.Audio.Media.EXTERNAL_CONTENT_URI,
						cursor.getLong(id)
					)
					if(tracks.any { it.uri == uri }) continue
					Track(
						context,
						uri,
						cursor.getString(data),
						cursor.getString(title),
						cursor.getString(artist),
						cursor.getString(album),
						cursor.getString(albumArtist),
						cursor.getString(cdTrackNumber),
						cursor.getString(discNumber)
					)
				}
			}
		} catch(e: Exception) {
			logE("Library.scanUri cursor: ${e.message}")
			error.emit(R.string.fail_scan_media)
		}
		try {
			//logD{"Library.scanUri sort"}
			tracks.sort()
			shuffleCounter = 0
			artists.sort()
			albums.sort()
			artists.forEach { it.sort() }
			albums.forEach { it.sort() }
			dirs.sort()
			dirs.forEach { it.shuffleCounter = 0 }
			Playlists.init(context)
		} catch(e: Exception) {
			logE("Library.scanUri sort: ${e.message}")
			error.emit(R.string.fail_scan_media)
		}
	}
	fun ensurePath(path: String): Int {
		val file = File("$musicDir$path")
		if(file.exists()) return CommsBT.CODE_FILE_EXISTS
		val parentFile: File = file.parentFile ?: return CommsBT.CODE_FAIL
		if(!parentFile.exists() && !parentFile.mkdirs()) return CommsBT.CODE_FAIL_CREATE_DIRECTORY
		if(!file.createNewFile()) return CommsBT.CODE_FAIL_CREATE_FILE
		return CommsBT.CODE_PENDING
	}
	fun deleteFile(main: Main, path: String): Int {
		val file = File("$musicDir$path")
		if(!file.exists()) {
			logI { "Library.deleteFile: path does not exists" }
			return CommsBT.CODE_OK
		}
		val uri = getUriForPath(path) ?: return CommsBT.CODE_FAIL
		if(main.checkUriPermission(
				uri,
				Process.myPid(),
				Process.myUid(),
				Intent.FLAG_GRANT_WRITE_URI_PERMISSION
			) == PackageManager.PERMISSION_DENIED
		) {
			try {
				main.deleteRequestResult.launch(
					IntentSenderRequest.Builder(
						MediaStore.createDeleteRequest(
							main.contentResolver,
							listOf<Uri?>(uri)
						).intentSender
					).build()
				)
			} catch(e: Exception) {
				logE("Library.deleteFile: startIntentSenderForResult: ${e.message}")
				return CommsBT.CODE_FAIL
			}
			return CommsBT.CODE_PENDING
		}
		if(!file.delete()) return CommsBT.CODE_FAIL_DEL_FILE
		scanFile(main, "$musicDir$path")
		return CommsBT.CODE_OK
	}
	private fun removeTrack(path: String) {
		tracks.removeIf { it.path == path }
		artists.forEach { artist ->
			artist.tracks.removeIf { t -> t.path == path }
			artist.albums.forEach { album ->
				album.tracks.removeIf { t -> t.path == path }
			}
			artist.albums.removeIf { it.tracks.isEmpty() }
		}
		artists.removeIf { it.tracks.isEmpty() }
		albums.forEach { album ->
			album.tracks.removeIf { t -> t.path == path }
		}
		albums.removeIf { it.tracks.isEmpty() }
		runInBackground { status.emit(Status.UPDATE) }
	}

	fun getTracksJson(): JSONArray {
		val array = JSONArray()
		tracks.forEach { array.put(it.toJson()) }
		return array
	}

	class Track(
		context: Context,
		val uri: Uri,
		path: String,
		var title: String, artistName: String?, albumName: String?, albumArtist: String?,
		trackNo: String?, discNo: String?
	) : Comparable<Track> {
		val path: String = path.substring(musicDir.length)
		val artist: Artist
		var album: Album? = null
		val trackNo: String?
		val discNo: String?

		init {
			this.title = title
			this.trackNo = trackNo
			this.discNo = discNo
			artist = getArtistForNewTrack(this, artistName)
			if(albumName != null && albumName != "Music") {
				album = getAlbumForNewTrack(context, this, albumName, albumArtist)
				artist.addAlbum(album as Album)
			}
			tracks.add(this)

			val dir = this.path.substringBeforeLast("/", "")
			if(dir.isNotEmpty()) {
				if(dirs.none { it.name == dir }) dirs.add(Dir(dir))
				dirs.first { it.name == dir }.tracks.add(this)
			}
		}
		fun toJson(): JSONObject {
			val ret = JSONObject()
			ret.put("path", path)
			return ret
		}
		override fun compareTo(other: Track): Int {
			var compare = humanCompare(album?.name, other.album?.name)
			if(compare != 0) return compare
			compare = humanCompare(discNo, other.discNo)
			if(compare != 0) return compare
			compare = humanCompare(trackNo, other.trackNo)
			if(compare != 0) return compare
			return humanCompare(title, other.title)
		}
	}

	class Artist(track: Track, artistName: String?) : Comparable<Artist> {
		val id: Int = artists.size
		val name: String = artistName ?: "<empty>"
		val tracks = mutableListOf<Track>()
		var shuffleCounter by mutableIntStateOf(0)
		val albums = mutableListOf<Album>()

		init {
			tracks.add(track)
			artists.add(this)
		}
		fun addAlbum(album: Album) {
			if(albums.contains(album)) return
			albums.add(album)
		}
		override fun compareTo(other: Artist): Int = name.compareTo(other.name, ignoreCase = true)
		fun sort() {
			tracks.sort()
			shuffleCounter = 0
			albums.sort()
		}
	}

	fun getArtistForNewTrack(track: Track, artistName: String?): Artist {
		val artist = artists.find { it.name == artistName }
		artist?.tracks?.add(track)
		return artist ?: Artist(track, artistName)
	}

	class Album(
		track: Track,
		albumName: String?,
		albumArtist: String?
	) : Comparable<Album> {
		val id: Int = albums.size
		val name: String = albumName ?: "<empty>"
		var artist: String = albumArtist ?: track.artist.name
		val tracks = mutableListOf<Track>()
		var shuffleCounter by mutableIntStateOf(0)

		init {
			tracks.add(track)
			albums.add(this)
		}
		override fun compareTo(other: Album): Int = name.compareTo(other.name, ignoreCase = true)
		fun sort() {
			tracks.sort()
			shuffleCounter = 0
		}
	}

	fun getAlbumForNewTrack(
		context: Context,
		track: Track,
		albumName: String?,
		albumArtist: String?
	): Album {
		return albums.find { it.name == albumName }?.apply {
			tracks.add(track)
			if(albumArtist != null && artist != albumArtist) {
				artist = context.getString(R.string.various_artists)
			}
		} ?: Album(track, albumName, albumArtist)
	}

	class Dir(val name: String) : Comparable<Dir> {
		val id: Int = dirs.size
		val tracks = mutableListOf<Track>()
		var shuffleCounter by mutableIntStateOf(0)
		override fun compareTo(other: Dir): Int = name.compareTo(other.name, ignoreCase = true)
	}

	private fun getUriForPath(path: String): Uri? = tracks.find { it.path == path }?.uri
	fun getFreeSpace() = File(exStorageDir).freeSpace

	fun humanCompare(value: String?, other: String?): Int {
		if(value == null && other == null) return 0
		if(value == null) return -1
		if(other == null) return 1
		val valueInt = value.toIntOrNull()
		val otherInt = other.toIntOrNull()
		if(valueInt == null && otherInt != null) return -1
		if(valueInt != null && otherInt == null) return 1
		if(valueInt != null && otherInt != null) return valueInt.compareTo(otherInt)
		val regex = Regex("(\\d+)|(\\D+)")
		val parts1 = regex.findAll(value).map { it.value }.toList()
		val parts2 = regex.findAll(other).map { it.value }.toList()
		for(i in 0 until minOf(parts1.size, parts2.size)) {
			val p1 = parts1[i]
			val p2 = parts2[i]
			val result =
				if(p1[0].isDigit() && p2[0].isDigit()) p1.toInt().compareTo(p2.toInt())
				else p1.compareTo(p2, ignoreCase = true)
			if(result != 0) return result
		}
		return parts1.size - parts2.size
	}
}